parity-clib/parity.h

```diff
@@ -86,21 +88,52 @@ int parity_start(const ParityParams* params, void** out);
+///
+///	- parity		: Reference to the running parity client
+///	- rpc_query		: JSON encoded string representing the RPC request.
+///	- len			: Length of the RPC query
+///	- timeout_ms	: Maximum time that request is waiting for a response
+///	- response		: Callback to invoke when the query gets answered. It will respond with
+///
+///						1) A JSON encoded string with the result
+///						2) A string "empty", (got an empty response)
+//						3) A string "timeout", (the query timed-out)
 ///
-/// Blocks the current thread until the request is finished. You are therefore encouraged to spawn
-/// a new thread for each RPC request that requires accessing the blockchain.
+/// - On success	: The parity client reference and the query string were valid
+/// - On error		: The parity client reference and the query string were not valid
 ///
-/// - `rpc` and `len` must contain the JSON string representing the RPC request.
-/// - `out_str` and `out_len` point to a buffer where the output JSON result will be stored. If the
-///	  buffer is not large enough, the function fails.
-/// - `out_len` will receive the final length of the string.
-/// - On success, the function returns 0. On failure, it returns 1.
+int parity_rpc(const void *const parity, const char* rpc_query, size_t rpc_len, size_t timeout_ms, callback response);
+
+
+/// Subscribes to the specified websocket event
+///
+///	 - parity		: Reference to the running parity client
+///	 - ws_query		: JSON encoded string representing the websocket and which event to subscribe to
+///	 - len			: Length of the queury
+///	 - response		: Callback to invoke when a websocket event occured
+///
+///	 - On success	: The function returns the underlying pointer to a atomic reference counter of the session object
+//					  which should later be used cancel the subscription
+///	 - On error		: The function returns a null pointer
+///
+const void *const parity_subscribe_ws(const void *const parity, const char* ws_query, size_t len, callback response);
+
+/// Unsubscribes from a specific websocket event. Caution this function consumes the session object and must only be
+/// exactly per session.
 ///
-/// **Important**: Keep in mind that this function doesn't write any null terminator on the output
-///                string.
+///	 - parity		: Reference to the running parity client
+///	 - session		: Underlying pointer to an atomic reference counter
+///	 - ws_query		: JSON encoded string representing the websocket event to unsubscribe from
+///	 - len			: Length of the query
+///	 - timeout		: Maximum time in milliseconds to wait for a response
+///	 - response		: Callback to invoke when a websocket event is received
 ///
-int parity_rpc(void* parity, const char* rpc, size_t len, char* out_str, size_t* out_len);
+///	 - On success	: The function return 0
+///	 - On error		: The function returns non-zero
+//
+int parity_unsubscribe_ws(const void *const parity, const void *const session, const char* ws_query,
```

Comment:
hrm, the `PubSubSession API` doesn't allows us to cancel subscription because these are private, https://github.com/paritytech/jsonrpc/blob/master/pubsub/src/subscription.rs#L32-#L73.

Thus, we need to use parity client to so. @tomusdrw can you confirm or suggest how do it?

Reply:
@niklasad1 It's enough to just drop the `Session` and it will unsubscribe automagically (of course if you want to unsubscribe from all subscriptions and you don't need to wait for that to finish).